Fixed #25389 -- Fixed pickling a SimpleLazyObject wrapping a model.

benjaminjkraft opened the pull request:
Pickling a `SimpleLazyObject` wrapping a model did not work correctly; in
particular it did not add the `_django_version` attribute added in 42736ac8.
Now it will handle this and other custom `__reduce__` methods correctly.

Also added a regression test.

MarkusH commented:
buildbot, test this please.

timgraham commented:
An alternate implementation seems to be: https://github.com/django/django/pull/5337. Would you mind discussing the best implementation with that author?

knifenomad commented:
I'm sure this fix is better way to go. I was worried about touching "__dict__" but this becomes another issue when @benjaminjkraft's fix is shipped. I'd rather create another ticket where my fix becomes necessary :)

timgraham commented on django/utils/functional.py:
```diff
@@ -247,32 +246,17 @@ def _setup(self):
-    # that pickle interprets this class.  This fails when the wrapped class is
-    # a builtin, but it is better than nothing.
-    def __getstate__(self):
+    # class we are pickling.  We're going to have to initialize the wrapped
```
You can standardize on 1 space after periods.

benjaminjkraft replied:
Sure, sorry about that.

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -19,6 +19,57 @@ def __eq__(self, other):
+    """
+    A base class with a funky __reduce__ method.
+
+    Meant to simulate the __reduce__ method of Model, which sets
```
Is this better than testing with an actual model? I imagine something in model could change in the future that could cause this to regress if we don't add a test for that.

benjaminjkraft replied:
No reason not to do both, I guess.

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -284,3 +335,17 @@ def test_list_set(self):
         self.assertEqual(len(lazy_list), 5)
         self.assertEqual(len(lazy_set), 4)
+
+    # Regression test for #25389
```
It might be better to create a separate test class so you can define the new objects right above it instead of having to reference the classes 300 lines above where the test is.

timgraham replied:
If you can describe the issue instead of simply "Regression test for #25389" that saves looking up the ticket to figure out the issue.

benjaminjkraft replied:
Sure.

timgraham commented:
Could you add release notes for 1.8.5? I'll admit, pickling isn't my expertise. Are you fairly confident this won't cause other regressions?

knifenomad commented:
This fix is nice if you plan to ship it on 1.8.5 because it solves the issue without messing with actual model class which could possibly result in pickling error due to change in model class while pickling.

benjaminjkraft commented:
I can do that.  I'm reasonably confident that this works in all the same cases as the old implementation, at least; it should be the same behavior in cases where `__reduce__` is not overridden, and the tests work in a number of configurations on each pickle version in both python 2 and python.  And I do think it's worth backporting if possible, because as @knifenomad's ticket points out, you don't have to do crazy things with `__dict__` to make this an error rather than just a warning or missing safety check -- you just have to have a `OneToOneField`, so this could be hit by any site that has a profile model attached to users and tries to pickle `request.user`.

Do you prefer that I push a second commit to this branch with the fix, force-push to this branch, or make a new branch with the fixes?

timgraham commented:
Force pushing the changes here is preferred. Thank you.

timgraham commented:
I'd like to release 1.8.5 this week, please let me know if you'll have time to update this patch for that release.

benjaminjkraft commented:
I looked at this a bit tonight, and ran into another issue.  I think I finally understand what's going on, but I need to think through the implications and potential fixes some more.  I'll take another look tomorrow evening.

benjaminjkraft commented:
Okay, I've pushed a fix to the issue I discovered while adding further tests.  (In particular, the case of a model with a `OneToOneField` was still raising the version check warning on unpickling.  This was roughly because to `pickle` it looked like there were two objects floating around with)  Now instead of trying to impersonate the wrapped object to be pickled, we simply use a custom unpickler with a single argument, namely the wrapped object, which returns that object.  This avoids the issue by having `pickle` pickle the wrapped object entirely normally (as an argument to the unpickler).

Pickling a `SimpleLazyObject` before this change and unpickling it after should work just fine.  (I tested this manually just to check, since writing an automated test for it would be tricky.)  The reverse (pickling an object after this change, and unpickling it before) won't work, but I think that's unavoidable in the bug fix.  If you really want to avoid that in a point release, I can make a version for 1.8.5 which does a partial fix (basically just the original version of this patch, which will fix the errors but will still cause warnings in some cases) and does not prevent sending pickles to old versions, and keep this in 1.9 and master.  It seems unnecessary to me, but I'm happy to do things either way.

That's the only potential issue I can think of.  I've added some extra tests using actual models, to better catch future regressions.  Lastly, I added a line to the release notes.  Let me know if there's anything else I can do!

timgraham commented on tests/utils_tests/models.py:
```diff
@@ -1,3 +1,4 @@
+from django.contrib.auth.models import User
```
I think it'd be better not to use models from other applications.

benjaminjkraft replied:
Done.

benjaminjkraft commented:
(fixed the lsort issue)

timgraham commented:
buildbot, add to whitelist.

timgraham commented on tests/utils_tests/models.py:
```diff
@@ -11,3 +11,11 @@ def next(self):
     category = models.ForeignKey(Category, models.CASCADE)
+
+
+class FooThing(models.Model):
```
any reason not to reuse `Category` instead of this new model? (Each model added makes the test suite a bit slower, so we try to reuse where possible -- not sure if the custom `next()` method that it has is an issue or not).

benjaminjkraft replied:
It shouldn't be an issue, I can reuse.  (Fixed that and the other issues.)

timgraham commented on django/utils/functional.py:
```diff
@@ -247,32 +246,30 @@ def _setup(self):
+
+    # We have to explicitly override __getstate__ so that older versions of
+    # pickle don't try to pickle the __dict__ (which in the case of a
+    # SimpleLazyObject may contain a lambda).  The value will end up being
```
chop extra space after period

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -284,3 +287,95 @@ def test_list_set(self):
+    # Test in a fairly synthetic setting
+    def test_pickle_with_reduce(self):
+        # Test every pickle protocol available
+        for protocol in range(pickle.HIGHEST_PROTOCOL):
```
`range(pickle.HIGHEST_PROTOCOL + 1)`

benjaminjkraft replied:
Good catch, thanks.

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -284,3 +287,95 @@ def test_list_set(self):
+        super(BaseBaz, self).__init__()
+
+
+# Regression test for issues with pickling a SimpleLazyObject wrapping a model
```
these comments above the test class/method should be docstrings I think

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -284,3 +287,95 @@ def test_list_set(self):
+            lazy_foo.barthing
+            lazy_foo_2 = SimpleLazyObject(lambda: foo)
+            with warnings.catch_warnings(record=True) as recorded:
+                self.assertEqual(
```
single line for these assertions is okay (we prefer longer lines if it helps readability)

timgraham commented on tests/utils_tests/test_lazyobject.py:
```diff
@@ -284,3 +287,95 @@ def test_list_set(self):
+        # Test every pickle protocol available
+        for protocol in range(pickle.HIGHEST_PROTOCOL):
+            lazy_foo = SimpleLazyObject(lambda: foo)
+            # Test it both if we accessed a field on the model, and if we
```
chop comma

timgraham commented:
merged in 35355a4ffedb2aeed52d5fe3034380ffc6a438db, thanks!

ionelmc commented on django/utils/functional.py:
```diff
@@ -311,6 +308,16 @@ def __deepcopy__(self, memo):
+    Actually just returns its argument, which will be the wrapped object.
+    """
+    return wrapped
+unpickle_lazyobject.__safe_for_unpickling__ = True
```
Note that this line is unnecessary. Checks for `__safe_for_unpickling__` were removed in Python 2.3: https://docs.python.org/2/whatsnew/2.3.html?highlight=__safe_for_unpickling__#pep-307-pickle-enhancements

timgraham replied:
Removed in 4accffd8fdaa59597dc40b47be031e7116c5b2a5, thanks.

ionelmc replied:
It's ea8e7fd989095c1444528c8b9808076985d5ea0a, but https://github.com/django/django/commit/4accffd8fdaa59597dc40b47be031e7116c5b2a5 is also quite interesting :-)
